Add `file_scan_inclusions` setting to customize Zed file indexing

Hawkbawk opened the pull request:
Closes #4745

Release Notes:

- Added a new `file_scan_inclusions` setting to force Zed to index files that match the provided globs, even if they're gitignored.

Hawkbawk committed:
Add support for always_included_files

Because there are some files (such as .env, or
docker-compose.override.yml, or others) that are ignored by git but are
still useful to search for in the file finder.

Hawkbawk commented:
Some things that I think are worth discussing:
1. Currently, if the setting is edited, we just force a complete reindex. Another possible approach here might be to keep a list of all always included entries then force a refresh on those entries when restarting the background scanner using something like `Worktree#refresh_entries_for_paths`
2. I'm not sure what defaults, if any, we'd like. I chose two, but those are based on my experience of what files I often find myself wanting to search that are typically ignored. Other folks will have other needs. It might be worth keeping the defaults as barebones as possible for now.
3. If someone provides a glob that begins with the globstar, for example `**/foo.txt`, Zed has to descend into every directory to see if there's a `foo.txt`. In particularly large repositories, this could lead to longer indexing times and results popping into the file finder as folks search. There's not really a way around this, as this is just how globs work, but I'm wondering if we should document this somehow?

Lemme know your thoughts, whoever ends up looking at this!

mikayla-maki commented on crates/worktree/src/worktree.rs:
```diff
@@ -3174,6 +3200,12 @@ pub struct Entry {
+    ///
+    /// This is used for entries that are always included in searches, even
+    /// if they are ignored by git or by `file_scan_exclusions`.
+    pub is_always_included: bool,
```
This isn't the approach we'd like to take for this feature. `Entry` should only be a representation of all of the local file system state we care about. Conceptually, `file_scan_inclusions` should be an entirely client-local concern, implemented as query on top of this state. We shouldn't need to involve collab or the database to implement this feature.

`file_scan_exlcusions` doesn't work this way because it's more of an optimization for common types of files. I think it's ok for `file_scan_exclusions` to trump `file_scan_inclusions`.

Hawkbawk replied:
Could you elaborate more on the approach of `client-local` concern? I'm not sure I totally understand. There needs to be some way to modify the indexing behavior of Zed to force entries to always be included in its internal representation of the filesystem.

Another way I could see doing this without needing to modify the database schema or send anything over the wire is just changing all of the current call sites from `entry.is_always_included` to something like `self.settings.is_path_always_included(path)`. Is that more along the lines of what you were thinking?

Thanks for taking a look and letting me know!

mikayla-maki replied:
What I mean by 'client local', is that this setting shouldn't need to be synced over the wire as an additional field on the `Entry`. Zed user Alice should read all the files as they do now (excluding files matching `file_scan_exclusions`), and then when rendering or otherwise querying this data structure, Alice should check this setting and whether it's git ignored. Meanwhile, Zed user Bob (who Alice is currently pairing with), should receive all of the `Entry`s, like we're doing now, and then when rendering or otherwise working with these files, it locally computes whether the file should be included by combining his local `file_scan_inclusions` with the git ignored state. 

So, having written that out, I think I am recommending the second approach you mentioned :)

Does that answer the question?

Hawkbawk replied:
Yes, that makes sense now. Thanks for clarifying! I'll go ahead and make those changes and let you know when this PR is ready for review again.

mikayla-maki replied:
Awesome :D just mark it as ready for review when you're ready.

Hawkbawk replied:
Alright, this should be good to go. I ended up having to either pass around the `WorktreeSettings` struct to some functions or add it as a field to some other structs, which I'm a bit reticent to do for some reason, but the database/collab should no longer be involved. Let me know if there's any other concerns!

mikayla-maki replied:
Great!

mikayla-maki commented on crates/worktree/src/worktree.rs:
```diff
@@ -987,11 +1000,35 @@ impl LocalWorktree {
         let (path_prefixes_to_scan_tx, path_prefixes_to_scan_rx) = channel::unbounded();
         self.scan_requests_tx = scan_requests_tx;
         self.path_prefixes_to_scan_tx = path_prefixes_to_scan_tx;
+        if force_reindex {
```
Why does changing our always_include settings require us to re-create the entire snapshot? Shouldn't we just queue the matching paths for re-scanning via the `scan_requests_tx`?

Hawkbawk replied:
Previous paths that were always_included will get left in the index, so if someone changes their setting, those old entries will still be left around in addition to the new ones that got added. An example might be if someone adds a directory to their always_included setting and realizes it's clogging up their search results (looking at you node_modules). Without the reindex, all of those entries will still be around until the user quits Zed and forces it to reindex anyways.
One other approach that I thought about but never tried to implement was keeping around a list of which entries are always_included, then using that list to remove obsolete entries whenever the setting changes. The reason I went with this approach instead was:
1. It was a lot easier :)
2. It seemed like this setting wouldn't get changed very often and the reindexing was fast enough that it wasn't a problem.

I bet there are almost certainly better ways to go about doing this though. This is definitely an area of the code that I feel the hackiest about.

mikayla-maki replied:
Fair, I think keeping the old paths around, and then merging the new paths with the old paths and sending them all for re-indexing is a preferable solution here. It scales better for super-large projects (e.g. the Linux source), and allows us to leave supporting code as it is. + We  can avoid having a boolean argument :)

Hawkbawk replied:
@mikayla-maki Alright, I believe this should be fixed now!

mikayla-maki commented on crates/worktree/Cargo.toml:
```diff
@@ -36,7 +36,7 @@ log.workspace = true
 paths.workspace = true
 postage.workspace = true
-rpc.workspace = true
+rpc = { workspace = true, features = ["gpui"]}
```
This seems like it shouldn't be here?

Hawkbawk replied:
Ah, this was to fix a weird thing I ran into while running individual tests locally. I'm hoping you've got a better solution in mind.
 The `worktree` crate relies on the `rpc::AnyProtoClient` struct, but as part of [this commit](https://github.com/zed-industries/zed/commit/01bb10f5186deb69db6120ac1572bed4fb69b55d), the proto_client stuff is only exported when `rpc` is compiled with the gpui feature. Any time I tried to run just a single worktree test, I'd get a complaint from the compiler about how the `AnyProtoClient` doesn't exist. I was able to run all tests just fine though, so I'm not sure if there's something wrong with how I was trying to run the tests (just a `cargo test -p worktree <test_name_here>`) or what.

mikayla-maki replied:
Ahhh, I see what the problem is. Feature unification fixes the compile errors, but only when you run the whole test suite...

mikayla-maki commented on crates/worktree/src/worktree.rs:
```diff
@@ -125,6 +125,7 @@ pub struct RemoteWorktree {
     background_snapshot: Arc<Mutex<(Snapshot, Vec<proto::UpdateWorktree>)>>,
     project_id: u64,
     client: AnyProtoClient,
+    settings: WorktreeSettings,
```
Can we just store the file_scan_inclusions here instead of all the settings?

Hawkbawk replied:
For sure! I think I was trying to get access to the info I needed and just decided to pass down the WorktreeSettings struct to get access to the `is_path_always_included` method, but it probably makes more sense to just pass the strictly needed info down.

mikayla-maki commented:
This is looking pretty good! Just a few implementation notes :)

Hawkbawk commented:
Thanks for the feedback and compliment! Responded!

Hawkbawk committed:
Fix Cargo.toml formatting

Hawkbawk committed:
Switch to better reindexing strategy

Rather than completely invalidate the snapshot when restarting the
background scanners, we can do better, and just make it reindex the
always_included entries to ensure they stay up-to-date.

ndiezel0 commented:
Hello. This change is critical for erlang-ls working in rebar3 projects in order to navigate dependencies. Is there some issue that needs to be resolved? I can help if it's needed.

mikayla-maki committed:
Merge branch 'main' into file_finder

mikayla-maki commented:
Thank you for implementing this!

notpeter commented on assets/settings/default.json:
```diff
@@ -679,6 +679,14 @@
+  // overly broad can slow down Zed's file scanning. Overridden by `file_scan_exclusions`.
+  "file_scan_inclusions": [
+    ".env*",
+    "docker-compose.*.yml"
```
Was intentional to include in the default settings?
I can understand `.env*`, but `docker-compose.*.yml` seems pretty niche.

Hawkbawk replied:
I think this was something that I wanted to be part of the discussion but it ended up being forgotten. We can definitely remove the `docker-compose.*.yml`. We should definitely try to keep this as minimal as possible to avoid adding any unnecessary and unexpected file scanning.

notpeter replied:
Fixed. Thanks.
